test/e2e/framework/util.go

```diff
@@ -4442,19 +4452,44 @@ const (
 	IPv6PingCommand PingCommand = "ping6"
 )
 
+// NcCommand is the type to hold nc command
+type NcCommand string
+
+const (
+	// IPv4NcCommand is a nc command for IPv4
+	IPv4NcCommand NcCommand = "nc"
+)
+
 // CheckConnectivityToHost launches a pod to test connectivity to the specified
 // host. An error will be returned if the host is not reachable from the pod.
 //
 // An empty nodeName will use the schedule to choose where the pod is executed.
-func CheckConnectivityToHost(f *Framework, nodeName, podName, host string, pingCmd PingCommand, timeout int) error {
+func CheckConnectivityToHost(f *Framework, nodeName, podName, host string, ipProto IPProtocol, timeout int) error {
```

Comment:
I don't think it makes sense for framework functions to have `IsProvider` checks. It makes reasoning about the e2e test unless you peak into the implementation details of `CheckConnectivityToHost`.  I think it would be cleaner if we just had two separate e2e tests, the existing one can just be skipped for Azure and the new one only runs for Azure (and possibly future providers that also block ICMP).

Reply:
With that said, I would also prefer to avoid `ProviderIs` checks in the e2e code base _if possible_, we're already trying to decouple provider code from e2e tests (https://github.com/kubernetes/kubernetes/issues/75604) and adding more checks like this adds more friction on that effort.

Reply:
@bowei What's your input on this? For #75604 I think it makes sense but I don't want to be blocked on that just yet. Once we have a solution for #75604 I can always follow the new pattern and change this.

Reply:
@andrewsykim Can you suggest a way to change this? This change is suggested by @bowei . I'd like to make the change after we all agree on the solution...

Reply:
That is a good point about keeping things provider agnostic in the framework if possible.

It would be a bigger effort, but ideally you would have a provider agnostic tools interface (e.g. with Ping(), etc) and the test would inject a provider specific "tools" struct to handle these callouts.
That would be the way I'd go, to keep things contained and easier to reason about.

Reply:
@bowei good idea, but it's still a long way to that. Another interface or framework is required.

Before that, what do you suggest for workarounds? `ProviderIs` is still the easiest way, right?

Reply:
@bowei I think we should bring your idea up in the next sig-testing/sig-network meeting and break the problem down. But getting that done may take weeks to months. I'd prefer to have a workaround for now and cleanup everything gradually once everybody gets assigned a piece of work. What do you think?

Reply:
The `ProviderIs` check is fine IMO, it's where we do the check that is problematic I think. It's cleaner to do to have two separate functions in this case, `CheckICMPConnectivityToHost` and `CheckTCPConnectivityToHost` and vary the call based on the provider from the test (`test/e2e/network/networking.go`) not the framework itself.

Reply:
@andrewsykim I agree. That's exactly what I was thinking when I did the first commit. I'll wait for @bowei and @feiskyer 's input on this.

Reply:
Why do we need a provider test?  Can we just switch to NC for everyone?

Reply:
That'd be the cleanest solution!! I just don't know if you guys would be OK with it. If so, I'll just switch to NC for everyone.

Reply:
The value of ping test is being able to validate that the LB is reachable regardless of the backend state. I see value in having both cases test but nc working would have overlapping validations with ping.

Reply:
I don't think we want to use `nc` because IIRC it does not exist for Windows hosts. We should use the agnhost image, proably, with either `curl` or a new Go-based command that makes an HTTP connection. This will avoid all platform and provider-specific issues.

Reply:
agnhost has both nc and curl commands